### PR TITLE
Calendar navigation

### PR DESCRIPTION
**Pull request by zarend**
Draft PR of fixing focus and navigation issues on the calendar. Please see commit messages for details on these changes.

This is still a draft at least until I can do more testing across different browsers and screen readers.

### CODE REVIEW

File: src/material/datepicker/calendar-body.html [hunk 1]
```diff
@@ -26,40 +26,59 @@
...
+    </td>
+
+    We render the calendar body cells as buttons to ensure that VoiceOver announces them as
+    interactable (issue #23476). A `gridcell` wraps each button, so that table-specific navigation
+    features are available to assistive technology. The gridcell is only used for semantic purposes
+    and does not affect the visual appearance.
+  -->
+  <td *ngFor="let item of row; let colIndex = index" role="gridcell" class="mat-calendar-body-cell-wrapper">
+    <div
+        role="button"
```
Comment (crisbeto) [inline src/material/datepicker/calendar-body.html:47]: Why don't we use a `button` element instead of a `div` with a `role`?

File: src/material/datepicker/calendar-body.scss [hunk 2]
```diff
@@ -31,7 +31,12 @@ $calendar-range-end-body-cell-size:
   padding-right: $calendar-body-label-side-padding;
 }
 
+.mat-calendar-body-cell-wrapper {
+  display: contents;
```
Comment (crisbeto) [inline src/material/datepicker/calendar-body.scss:35]: Why is this necessary? Also it seems this isn't supported in Safari.

File: src/material/datepicker/calendar-body.scss [hunk 3]
```diff
@@ -31,7 +31,12 @@ $calendar-range-end-body-cell-size:
...
+}
+
 .mat-calendar-body-cell {
+  display: table-cell;
```
Comment (crisbeto) [inline src/material/datepicker/calendar-body.scss:39]: We had an issue where NVDA would announce an extra "table" because something had `display: table`. I wouldn't be surprised if it becomes an issue here too. See https://github.com/angular/components/pull/23446.

File: src/material/datepicker/calendar-body.ts [hunk 4]
```diff
@@ -122,6 +122,9 @@ export class MatCalendarBody implements OnChanges, OnDestroy {
   /** Emits when a new value is selected. */
   @Output() readonly selectedValueChange = new EventEmitter<MatCalendarUserEvent<number>>();
 
+  /** Emits when a new date becomes active. */
+  @Output() readonly activeValueChange = new EventEmitter<MatCalendarUserEvent<number>>();
```
Comment (crisbeto) [inline src/material/datepicker/calendar-body.ts:126]: What is this event used for? It seems like we're only emitting it/forwarding it.

File: src/material/datepicker/calendar-body.scss (hunk 2 above)
Comment (zarend) [reply src/material/datepicker/calendar-body.scss:35]: Hmm, well, the [MDN Compatibility table for display](https://developer.mozilla.org/en-US/docs/Web/CSS/display#browser_compatibility) says there is some issues on Safari with unusual elements 🤔 . I tried this on Safari and it seems to render fine.


Safari Version 15.1 (17612.2.9.1.20)

File: src/material/datepicker/calendar-body.html (hunk 1 above)
Comment (zarend) [reply src/material/datepicker/calendar-body.html:47]: For the draft implementation of it, it was the path of least resistance to use `div role="button"` because it doesn't affect the visual styles. We would need to override a few (not sure exactly how many) style rules to get a `button` element to look correct.

File: src/material/datepicker/calendar-body.ts (hunk 4 above)
Comment (zarend) [reply src/material/datepicker/calendar-body.ts:126]: Hmm, good question. What this is trying to accomplish is making it so that the calendar body can update the value of the `activeDate` when a cell receives the focus event. `activeDate` has a reference in three components – the root component MatCalendarBody, the component for the specific view month/year/multiyear and the calendar body.

This might be easier to do with banana in a box rather than passing an event. That is currently (in master branch) how the `MatCalendar` component gives the `activeDate` to the MatMonthView component (and also MatMonthView and MatMultiYearView).

https://github.com/angular/components/blob/03485cd63e7622da839c115679f261b15989e33c/src/material/datepicker/calendar.html#L6

File: src/material/datepicker/calendar-body.html (hunk 1 above)
Comment (crisbeto) [reply src/material/datepicker/calendar-body.html:47]: It's only 5-6 lines of CSS so it might be easier to do it now https://github.com/angular/components/blob/master/src/material/core/style/_button-common.scss#L5.

File: src/material/datepicker/month-view.spec.ts [hunk 5]
```diff
@@ -149,13 +149,14 @@ describe('MatMonthView', () => {
             fixture.detectChanges();
           });
 
-          it('should decrement date on left arrow press', () => {
+          fit('should decrement date on left arrow press', () => {
```
Comment (crisbeto) [inline src/material/datepicker/month-view.spec.ts:152]: Leftover from debugging?

File: src/material/datepicker/calendar-body.ts (hunk 4 above)
Comment (crisbeto) [reply src/material/datepicker/calendar-body.ts:126]: ok, so if I'm understanding it correctly, what it's trying to achieve is to sync the `activeDate` back up with the calendar if focus is moved by something we don't control (e.g. a keyboard shortcut)?

Comment (zarend) [reply src/material/datepicker/calendar-body.ts:126]: Exactly, if I had more time, I would even go as far as renaming `activeDate` to `focusedDate`, but I'm not sure that would be worth changing public API for.

**Comment by zarend** [comment]
@crisbeto thanks for looking this draft. Over the break, it was easier to work on multiple changes on the same branch until we started merging PR's again.

I'm closing this in favor of individual non-draft PR's for these changes. Please see:
 - #24168
 - #24171
 - #24172
